client/src/components/ElementEditor/Element.js

```diff
@@ -215,6 +302,102 @@ class Element extends Component {
+      justClickedPublishButton: true,
+      ensureFormRendered: true,
+    });
+  }
+
+  handleAfterSave() {
+    this.setState({
+      doSaveElement: false,
+    });
+  }
+
+  handleAfterPublish(wasError) {
+    this.showPublishedElementToast(wasError);
+    this.setState({
+      doPublishElement: false,
+      doPublishElementAfterSave: false,
+    });
+    // Ensure that formDirty becomes falsey after publishing
+    // We need to call at a later render rather than straight away or redux-form may override this
+    // and set the form state to dirty under certain conditions
+    // setTimeout is a hackish way to do this, though I'm not sure how else we can do this
+    // The core issue is that redux-form will detect changes when a form is hydrated for the first
+    // time under certain conditions, specifically during a behat test when trying to publish a closed
+    // block when presumably the apollo cache is empty (or something like that). This happens late and
+    // there are no hooks/callbacks available after this happens the input onchange handlers are fired
+    setTimeout(() => {
+      this.props.dispatchRemoveFormChanged();
+    }, 500);
+    this.refetchElementalArea();
+  }
+
+  handleFormInit(activeTab) {
+    this.updateFormTab(activeTab);
+    this.setState({
+      formHasRendered: true,
+    });
+  }
+
+  handleFormSchemaSubmitResponse(formSchema, title) {
+    // Slightly annoyingly, on validation error formSchema at this point will not have an errors node
+    // Instead it will have the original formSchema id used for the GET request to get the formSchema i.e.
+    // admin/elemental-area/schema/<ItemID>
+    // Instead of the one used by the POST submission i.e.
+    // admin/elemental-area/elementForm/<LinkID>
+    const hasValidationErrors = formSchema.id.match(/\/schema\/elemental-area\/([0-9]+)/);
+    if (hasValidationErrors) {
+      if (this.props.type.inlineEditable) {
+        this.setState({
+          previewExpanded: true,
+        });
+      }
+      // Ensure that formDirty remains truthy
+      // Note we need to call this.props.dispatchAddFormChanged() on the next render rather than straight away
+      // or it will get unset by code somewhere else, probably redux-form
+      this.setState({
+        doDispatchAddFormChanged: true,
+      });
+      // Don't accidentally auto publish the element once validation errors are fixed
+      if (this.state.doPublishElementAfterSave) {
+        this.setState({
+          doPublishElementAfterSave: false,
+        });
+      }
+      return;
+    }
+    // Form is valid
+    this.setState({
+      newTitle: title,
+    });
+    if (this.state.doPublishElementAfterSave) {
+      this.setState({
+        doPublishElementAfterSave: false,
+        doPublishElement: true,
+      });
+    }
+    if (!this.state.doPublishElement && !this.state.doPublishElementAfterSave) {
+      this.showSavedElementToast(title);
+    }
+    this.refetchElementalArea();
```

Comment:
```suggestion
    const newState = {};
    if (hasValidationErrors) {
      if (this.props.type.inlineEditable) {
        newState.previewExpanded = true;
      }
      // Ensure that formDirty remains truthy
      // Note we need to call this.props.dispatchAddFormChanged() on the next render rather than straight away
      // or it will get unset by code somewhere else, probably redux-form
      newState.doDispatchAddFormChanged = true;
      // Don't accidentally auto publish the element once validation errors are fixed
      if (this.state.doPublishElementAfterSave) {
        newState.doPublishElementAfterSave = false;
      }
      this.setState(newState);
      return;
    }
    // Form is valid
    newState.newTitle = title;
    if (this.state.doPublishElementAfterSave) {
      newState.doPublishElementAfterSave = false;
      newState.doPublishElement = true;
    }
    if (!this.state.doPublishElement && !this.state.doPublishElementAfterSave) {
      this.showSavedElementToast(title);
    }
    this.setState(newState);
    this.refetchElementalArea();
```
As above, I suspect we don't need to be making so many `setState()` calls.

Reply:
Updated